pallets/gear-voucher/src/internal.rs

```diff
@@ -0,0 +1,202 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![allow(unused)]
+
+use crate::*;
+use common::{
+    storage::{Counter, CounterImpl, Mailbox},
+    DelegateFee, Origin,
+};
+use gear_core::{declare_id, ids};
+use sp_std::collections::btree_set::BTreeSet;
+
+declare_id!(VoucherId: "Voucher identifier");
+
+impl VoucherId {
+    pub fn generate<T: Config>() -> Self {
+        const SALT: &[u8] = b"voucher";
+
+        let nonce = CounterImpl::<u64, IssuedWrap<T>>::inc_get();
+
+        let argument = [SALT, &nonce.to_le_bytes()].concat();
+        ids::hash(&argument).into()
+    }
+}
+
+impl Origin for VoucherId {
+    fn into_origin(self) -> H256 {
+        self.0.into()
+    }
+
+    fn from_origin(val: H256) -> Self {
+        Self(val.to_fixed_bytes())
+    }
+}
+
+/// Type containing all data about voucher.
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub struct VoucherInfo<AccountId, BlockNumber> {
+    /// Owner of the voucher.
+    /// May be different to original issuer.
+    /// Owner manages and claims back remaining balance of the voucher.
+    pub owner: AccountId,
+    /// Set of programs this voucher could be used to interact with.
+    /// In case of None means any gear program.
+    pub programs: Option<BTreeSet<ProgramId>>,
+    /// Block number since voucher couldn't be used (able to be revoked by owner).
+    pub validity: BlockNumber,
+}
+
+impl<AccountId, BlockNumber> VoucherInfo<AccountId, BlockNumber> {
+    pub fn contains(&self, program_id: ProgramId) -> bool {
+        self.programs
+            .as_ref()
+            .map(|v| v.contains(&program_id))
+            .unwrap_or(true)
+    }
+}
+
+/// Prepaid call to be executed on-chain.
+#[derive(Debug, Clone, Encode, Decode, TypeInfo, PartialEq, Eq, PartialOrd, Ord)]
+pub enum PrepaidCall<Balance> {
```

Comment:
Couldn't we get away without this artificial type? It only serves for the purpose of validating a random call eligibility for being "prepaid". We could pass a usual call filter associated type that would filter out illegitimate calls, could we? Which, in my opinion, would be a more "Substrate-idiomatic" way. Or am I missing something?

Reply:
This way there would be no need for a custom call dispatcher.

Reply:
Yeah, but it seems weird to accept the type of call where 98% are invalid and couldn't be used with voucher logic.
With approach of separate enum type we
1) increase ui for polkajs and similars;
2) specify exact calls that could be processed (instead of misleading to all of them)
3) (:c) have overhead of a custom call dispatcher, but it seems fine for me

Furthermore, it's considering for future to issue vouchers for some specific actions: e.g. only send_messages or anything else: there will be problems parsing runtime call every time (or implementing try into manually)

Reply:
Regarding the last argument, if we look at the Substrate's own `proxy` pallet, we'll see quite a similar use case: there they have a hierarchy of `ProxyTypes`, and filter calls based on what a specific proxy can or cannot do. It looks a bit complicated, I agree, but that's what they do.
Anyway, if you think your approach is better (apart from being more readable), then I'm fine.

Reply:
The topic is always open and I think we can change the solution implementation a bit later. Atm this is how ecosystem images their interactions with vouchers